### DISCUSSION TITLE
regex for any unicode character greater than 0x9F

### DISCUSSION THREAD

**Discussion by will** [tree-sitter/tree-sitter#1214] [Q&A] [answered]
Hi, I'm trying to write a regex that satisfies the subject, but running into troubles.

I first tried wrongly `[\u009F-\uFFFF]` but that is (I think) doing individual c chars, and not full unicode characters, so some things weren’t matching that I wanted to, like random emojis that take up several bytes.

I then tried making the range be `\u009F-\x{10FFFF}` and got `Invalid regular expression: ... Range out of order in character class`

Does anyone know how to make a regex that matches any unicode character greater than 0x9F?

edit: Also it doesn’t _have_ to be a regex. If there is some other way to do this with the grammar stuff that'd be cool too. Failing that I could start an external parser, but I'm hoping to put that off as far as I can.

**Comment by razzeee** [comment]
Not sure what your trying to do, but maybe there is a unicode category that matches? Tree-sitter started supporting these a few versions ago.
https://www.regular-expressions.info/unicode.html#category

**Reply by will** [reply]
> Not sure what your trying to do,

I'm looking into writing a grammar for Crystal. Symbols [are defined as](https://crystal-lang.org/reference/syntax_and_semantics/literals/symbol.html) and is a colon followed by "alphanumeric characters, underscore (_) or characters with a code point greater than 159(0x9F)." (for non quoted symbols at least)

Thanks for the link to the unicode categories, but it doesn't appear if any of them match that range.

**Reply by maxbrunsfeld** [reply]
I'm actually surprised that ranges like `[\u009F-\uFFFF]` don't work. They aren't byte-based - they're tested against the unicode code point. What behavior are you seeing with Tree-sitter when using those?

**Reply by will** [reply]
I'm very new to this, so it's quite possible I'm doing something wrong 😂

So I have this so far

``` javascript
  rules: {
    program: $ => repeat($._statement),

    _statement: $ => seq(
      choice(
        $._expression
      ),
      choice(';', /\n/),
    ),

    _expression: $ => choice (
      $.symbol                  ,
      $.string                  ,
    ),

    symbol: $ => seq(':', choice(
      /[a-zA-Z\u009F-\uFFFF][a-zA-Z0-9\u009F-\uFFFF_]*[=!\?]?/,
      $._operator,
      seq('"', repeat(choice(/[^"]/, '\"')), '"'),
    )),

    string: $ => seq('"', repeat(choice(/[^"]/, '\"')), '"'),

    _primary: $ => choice(
      $.symbol
    ),

    _operator: $ => choice( "+", "-", "*", "/", "%", "&", "|", "^", "**", ">>", "<<", "==", "!=", "<", "<=", ">", ">=", "<=>", "===", "[]", "[]?", "[]=", "!", "~", "!~", "=~",),
  }
```

and these all work 

```
:a
:a=
:unquoted_symbol
:question?
:exclamation!
:あ
:"ok yes!!"
:"this one has \" a quote"
:<=>
:[]?
```

However 

```
======
test
======

:🦉❓

---

(program (symbol))

```

is valid in the language (though silly)

```
~> crystal eval 'p :🦉❓'
:"🦉❓"
```

and with the grammar as above, I get 

```
    (program (symbol (ERROR (UNEXPECTED 129417)))) (symbol))
```

**Reply by maxbrunsfeld** [reply]
It looks like [the owl](https://emojipedia.org/emoji/%F0%9F%A6%89/)'s code point is `1F989`, but your range only goes up to `FFFF`. Maybe you need a different range? Unicode code points are 32-bit integers, so I believe the maximum possible value is `FFFFFFFF`.

**Reply by will** [reply]
Thanks. Trying these ways I got the `Range out of order in character class` error. There is maybe a way to figure out what ranges trip that error and carve them out of the regex, but @CyberShadow's negation idea is working.

**Comment by CyberShadow** [comment, marked as answer]
Sorry if this isn't helpful but have you tried simply negating the range to exclude any character not in the 00-9F range? I.e. `/[^\x00-\x9F]/`

**Reply by will** [reply]
Nice, thanks. I hadn't thought of negating. It's a little more complicated since there are a few other ranges that I do want to match, but I do think I've successfully negated the whole thing.

This is ugly but seems to work

```javascript
//        A-Z        _      a-z                  0-9      A-Z       _        a-z
/[^\x00-\x40\x5B-\x5E\x60-\x60\x7B-\x9F][^\x00-\x2F\x3A-\x40\x5B-\x5E\x60-\x60\x7B-\x9F]*[=!\?]?/,

```
